param_passing/proposal.md

```diff
@@ -0,0 +1,194 @@
+    <sensor name="lidar">
+      ...
+    </sensor>
+    <sensor name="camera">
+      <pose>0.5 0.02 0 0 0</pose>
+      ...
+    </sensor>
+  </link>
+
+  <link name="top">
+    <visual name="camera_visual">
+      ...
+      <geometry>
+        <box>
+          <size>0.03 0.03 0.05</size>
+        </box>
+      </geometry>
+      <material>
+        <ambient>0.0 1.0 0.0 1</ambient>
+        <diffuse>0.0 1.0 0.0 1</diffuse>
+        <specular>0.5 0.5 0.5 1</specular>
+      </material>
+    </visual>
+    <sensor name="camera">
+      <pose>0.2 1 0.5 0 0 0</pose>
+      <update_rate>20</update_rate>
+      ...
+    </sensor>
+  </link>
+```
+
+Now the user would like to use the original model `base_robot` in a world file but with modifications. The modifications include removing the lidar from the chassis link, adding a plugin to the camera sensor in the chassis link, adding a camera visual to the chassis link, update camera information in the top link, and replace the top link's camera visual geometry.
+
+```xml
+<model name="custom_robot">
+  <include>
+    <uri>model://base_robot</uri>
+
+    <experimental:params>
+      <visual name="chassis::lidar_visual" action="remove"/>
+      <sensor name="chassis::lidar" action="remove"/>
+
+     <sensor name="chassis::camera" action="add">
+        <plugin name="camera_depth_sensor" filename="libcamera_depth_sensor.so"/>
+     </sensor>
+
+      <visual name="chassis::camera_visual" action="add">
+        <pose>0.5 0.02 0 0 0</pose>
+        <geometry>
+            <box>
+                <size>0.02 0.02 0.02</size>
+            </box>
+        </geometry>
+      </visual>
+
+      <sensor name="top::camera" action="modify">
+        <pose>0 1 0 0 0 0</pose>
+        <update_rate>60</update_rate>
+      </sensor>
+
+      <visual name="top::camera_visual">
+        <geometry action="replace">
+          <sphere>
+            <radius>0.05</radius>
+          </sphere>
+        </geometry>
+        <material action="modify">
+          <ambient>0.0 1.0 0.0 1</ambient>
+        </material>
+      </visual>
+
+    </experimental:params>
+  </include>
+</model>
+```
+
+In `//include/experimental::params`, the action can be either listed in the element identifier or listed individually in the children element(s). For example,
+
+```xml
+<sensor name="top::camera" action="modify">
+  <pose>0 1 0 0 0 0</pose>
+  <update_rate>60</update_rate>
+</sensor>
+```
+
+The action is defined at the top level where the `name` identifier is, which means that all children listed in this element will be modified. If the user wants to do several different actions to the children elements then the user can specify the action for each child element individually. For instance,
+
+```xml
+<visual name="top::camera_visual">
+  <geometry action="replace">
+    <sphere>
+      <radius>0.05</radius>
+    </sphere>
+  </geometry>
+  <material action="modify">
+    <ambient>0.0 1.0 0.0 1</ambient>
+  </material>
+</visual>
```

Comment:
Can you give an explicit example of modifying an element without a name? For example, how would one modify just the radius of a sphere?

Reply:
See comment above https://github.com/osrf/sdf_tutorials/pull/46#discussion_r556732057. If we go with this route, then if the user wants to modify the radius of a sphere, then they would need to use `action="replace"`.

Reply:
After some thought, users should be able to use `action=modify` technically as well. Both of these examples would yield the same result:

```xml
<visual name="top::camera_visual">
  <geometry action="replace">
    <sphere>
       <radius>0.05</radius>
    </sphere>
  </geometry>
</visual>
```
```xml
<visual name="top::camera_visual">
  <geometry action="modify">
    <sphere>
       <radius>0.05</radius>
    </sphere>
  </geometry>
</visual>
```
The code that actually implements this would be different. replace would look for the //geometry tag and replace it with what's inside. What are you're thoughts? Should I add this example?

Reply:
Spoke offline with @azeey and the user would be able to use the action `modify` as well. Both of these examples would yield the same result:

```xml
<visual name="top::camera_visual">
  <geometry action="replace">
    <sphere>
       <radius>0.05</radius>
    </sphere>
  </geometry>
</visual>
```

```xml
<visual name="top::camera_visual">
  <geometry action="modify">
    <sphere>
       <radius>0.05</radius>
    </sphere>
  </geometry>
</visual>
```

The code that actually implements this would be different. `replace` would look for the `//geometry` tag and replace it with what's inside. Whereas `modify` would look through the children, compare them to the original to find what needs modified.

Added `modify` example added in 30237e6c33707a61c172b2551cbf7ddbb31208ed since it is more semantically correct.

Reply:
Thanks!